### ISSUE TITLE
feature: Auth component

### ISSUE THREAD

**Issue by fwang** [sst/ion#90]
Requests:
- https://discord.com/channels/983865673656705025/1177071497974648952/1222581598831120385

**Comment by beeirl** [comment]
Would be great to make the current adapters a bit more flexible:

[CodeAdapter](https://github.com/sst/ion/blob/dev/sdk/js/src/auth/adapter/code.ts#L55C7-L55C57)
Proper error handling for expired `authorization` cookie i.e.
```diff
- if (!authorization) throw new UnknownStateError();
+ if (!authorization) return ctx.forward(c, await config.onCodeExpired(c.req.raw))
```

[OauthAdapter](https://github.com/sst/ion/blob/dev/sdk/js/src/auth/adapter/oauth.ts)
Make the underlying `openid-client` configurable for custom oauth providers that require a different config.

Happy to open a PR but not sure if it's worth because I am assuming that the implementation might change anyway once you fully release that component.

**Comment by Meeeeez** [comment]
When will this be updated to v3?

**Comment by klclee** [comment]
ok did a digging around and found this: https://github.com/sst/ion/blob/dev/sdk/js/src/auth/adapter/oauth.ts

Tried to set it up using this example as reference: https://github.com/sst/ion/blob/dev/examples/aws-auth/src/auth.ts

event tho i have adapted the example to use the oauth params (very similar with v2 one). When I try to use the adapter to auth by navigating to the generated auth URL.  I am getting 'Missing redirect_uri'

**Comment by ryan-hunter-pc** [comment]
@klclee I was able to get auth working in v3 using a Google OIDC connector.

## TL;DR
[Here's a gist where I show how I got it working on my end](https://gist.github.com/ryan-hunter-pc/88e5ecf01cc8cd4c54637361e467be25)

If you're on the SST Discord, I posted [this message which briefly explains on the confusion that many are hitting](https://discord.com/channels/983865673656705025/1289642513258381344/1293327458740998214) (including me, initially).

## Full explanation, if you care...

_SST team, forgive me if this terminology clashes which how you plan to document it, but this is the mental model that helped me figure out how to replicate the slick v2 auth flow using the v3 components that exist as of now. Feel free to tell me to revise/remove this if is clashes with your intentions_

The handler created by calling `auth.authorizer(...)` (in v3) essentially provides an "outer" OAuth flow which uses whatever adapter(s) you want internally, which expects your app to follow a typical authorization code flow to get back a custom JWT session token.

The adapter(s) may then perform an "inner" OAuth/OIDC flow depending on the adapter, but that is separate from the "outer" flow provided by the SST authorizer. The confusion I'm seeing around Discord and here seems to be in not understanding the difference there.

v2 auth abstracted all that away nicely by providing the `auth.attach(...)` method which gave you a backend proxy to handle that "outer" authorization code flow, and then you could use the `Session.parameter(...)` or `Session.cookie(...)` methods to pass the resulting custom JWT session token back to your frontend however you wished.

Since v3 does not (yet?) have those abstractions in place, I believe the correct approach is to build your own "outer" flow which uses the authorizer created by `auth.authorizer(...)`. In my case, I created a backend proxy similar to what v2 provided via `auth.attach(...)`, but you could do it any number of ways.

The confusion is that some people (including me, at first) are just trying to link to `<your-authorizer-url>/google/authorize` directly from the frontend without any parameters, assuming that it'll handle the authorization code proxying for you like `auth.attach(...)` did in v2. That's what causes the `Missing redirect_uri` error, because the authorizer is expecting the parameters needed to kick off an authorization code flow.

My gist linked above is a bit hacky in how it wires things up, but it is (I believe) correctly following the "outer" and "inner" proxying, and safely isolates all auth knowledge to the backend. The result is that (just like in v2), you can link to `https://your-backend-api.com/auth/google` from your frontend, and there is a proxy handler which will forward that to the authorizer with the proper parameters, perform the authorization code flow, and then callback to your frontend URL with `?token=<token>` where `<token>` is the custom JWT session defined by `createSessionBuilder()`.
